[c10d][fr] Enable FR analysis script for rest of all coalesce op

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151238
* __->__ #151247
* #151243

We revisited how coalesced collective is working in https://github.com/pytorch/pytorch/pull/151243 and we now want to enable the script to work for slow path. The change is indeed bc-breaking but this is needed to make it work and the API is an internal use API. It is not user facing. For slow path the individual has input-sizes and output sizes recorded but no state. The final one has the state ready. We check the correctness of each individual collective one by one but we don't check the state match for these collectives, we can only check the state match for the last one which is the work item with coalesced label.

Added more unit test for slow path.


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[c10d][fr] Enable FR analysis script for all coalesce op

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Enable FR analysis script for rest of all coalesce op"


We revisited how coalesced collective is working in https://github.com/pytorch/pytorch/pull/151243 and we now want to enable the script to work for slow path. The change is indeed bc-breaking but this is needed to make it work and the API is an internal use API. It is not user facing. For slow path the individual has input-sizes and output sizes recorded but no state. The final one has the state ready. We check the correctness of each individual collective one by one but we don't check the state match for these collectives, we can only check the state match for the last one which is the work item with coalesced label.

Added more unit test for slow path.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Enable FR analysis script for rest of all coalesce op"


We revisited how coalesced collective is working in https://github.com/pytorch/pytorch/pull/151243 and we now want to enable the script to work for slow path. The change is indeed bc-breaking but this is needed to make it work and the API is an internal use API. It is not user facing. For slow path the individual has input-sizes and output sizes recorded but no state. The final one has the state ready. We check the correctness of each individual collective one by one but we don't check the state match for these collectives, we can only check the state match for the last one which is the work item with coalesced label.

Added more unit test for slow path.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

d4l3k commented on tools/flight_recorder/components/utils.py:
```diff
@@ -364,7 +445,8 @@ def find_coalesced_group(
     if len(found) > 1:
-        assert found[-1][1]["profiling_name"] == "nccl:coalesced"
+        if found[-1][1]["profiling_name"] != "nccl:coalesced":
+            logger.error("Rank %s does not have a coalesced end.", rank)
```
We're changing this so we can throw errors on mismatched coalesced operations?

fduwjj replied:
Right. If we throw exception, then we won't run to the following analysis part.

d4l3k reviewed (approved):
LGTM to unblock, I am a bit concerned about the BC changes -- we should have a discussion about the rollout strategy to avoid breaking existing users

I don't have much context on this code

fduwjj commented:
@d4l3k sure, I think that's a fair point. Although the API is just used internally by the analysis script. For now I will first merge https://github.com/pytorch/pytorch/pull/151243 since it is most critical and not bc-breaking. And this PR needs to work with https://github.com/pytorch/pytorch/pull/151238. Since we bump up FR version in https://github.com/pytorch/pytorch/pull/151238, what I will do is to use version check to call different function and keep the old function so that we don't even have bc-breaking for API.

wconstab commented:
I'll have to spend some more time to review this script change. can you say what the BC change is? Changing the behavior of an internal-only api isn't really 'BC-breaking' unless it's an api that users are depending on despite the underscore. If the script still works the same for existing users I wouldn't object to a behavior change in its helper functions.

XilunWu reviewed (approved):
Trust Tristan's review, stamp to unblock!

fduwjj commented:
@wconstab it is a internal api change but the linter still detects it to be bc-breaking..

fduwjj committed:
Update on "[c10d][fr] Enable FR analysis script for rest of all coalesce op"


We revisited how coalesced collective is working in https://github.com/pytorch/pytorch/pull/151243 and we now want to enable the script to work for slow path. The change is indeed bc-breaking but this is needed to make it work and the API is an internal use API. It is not user facing. For slow path the individual has input-sizes and output sizes recorded but no state. The final one has the state ready. We check the correctness of each individual collective one by one but we don't check the state match for these collectives, we can only check the state match for the last one which is the work item with coalesced label.

Added more unit test for slow path.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj committed:
Update on "[c10d][fr] Enable FR analysis script for rest of all coalesce op"


We revisited how coalesced collective is working in https://github.com/pytorch/pytorch/pull/151243 and we now want to enable the script to work for slow path. The change is indeed bc-breaking but this is needed to make it work and the API is an internal use API. It is not user facing. For slow path the individual has input-sizes and output sizes recorded but no state. The final one has the state ready. We check the correctness of each individual collective one by one but we don't check the state match for these collectives, we can only check the state match for the last one which is the work item with coalesced label.

Added more unit test for slow path.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
